ZENKO-3254: added two procedures to install and upgrade XDM

thomasj-tech opened the pull request:
<!--
Thank you for contributing to Zenko!

Please enter applicable information below.
-->

**What does this PR do, and why do we need it?**

**Which issue does this PR fix?**

Added two procedures:
- "Installing XDM" in Zenko Installation,
- "Upgrading XDM" in Zenko Operation.

fixes #<ISSUE>

**Special notes for your reviewers**:

thomasj-tech committed:
ZENKO-3254: added procedure to install XDM with MetalK8s

bert-e commented:
# Hello thomasj-tech,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/zenko/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-3254 contains:


* `2.0.0-alpha.4`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.0.0`


Please check the `Fix Version/s` of ZENKO-3254, or the target
branch of this pull request.

lucieleonard commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+  +-------------+--------+-----------+
+  | Partition # |  Size  |  Service  |
+  +=============+========+===========+
+  | 1           | 300GiB | Mongo     |
```
space between 300 and GiB (to apply everywhere)

lucieleonard commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+Deploying MetalK8s
+------------------
+
+For more information on MetalK8s installation, see *Installation* in the MetalK8s' documentation.
```
does not sound very good... Try to reword it :) 
You can avoid repeating MetalK8s and installation (hint: change "For more information about MetalK8s installation"). I'll let you think about it, if you cannot find anything let's talk about it

wabernat replied:
Install MetalK8s as described in its installation documentation.

wabernat replied:
If you want to italicize it, set it as "as described in *MetalK8s Installation*"

lucieleonard commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+Deploying XDM-Operator
+----------------------
+
+#. From your terminal, create a namespace for XDM with:
```
too many  "with:"
From the terminal, create a namespace for XDM. Simpler and better :)

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -12,21 +12,21 @@ Prerequisites
 
 - Your disk must be partioned as follows:
```
Partition your disk as follows:

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -12,21 +12,21 @@ Prerequisites
+  +-------------+---------+-----------+
+  | 4           | 10 GiB  | Redis     |
+  +-------------+---------+-----------+
+  | 5           | 10 GiB  | Zookeeper |
```
ZooKeeper

wabernat commented on docs/docsource/installation/install/index.rst:
```diff
@@ -11,4 +11,5 @@ configured, you must install Helm, then use it to install |product|.
 
    install_helm
    install_zenko
+   installing_xdm
```
I am concerned about parallelism here. I recommend changing the file name and this link to "install_xdm" as with Helm and Zenko, above.

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+.. _Installing XDM:
+
+Installing XDM
```
Install XDM (consistent with my previous comment)

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+
+For more information on MetalK8s installation, see *Installation* in the MetalK8s' documentation.
+
+Deploying XDM-Operator
```
That hyphen is up to no good. 

Deploy XDM Operator

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+      kubectl -n zenko rollout status --timeout 10m deploy kubedb-operator
+      sed "s/SOLUTION_ENV/zenko/g" /srv/scality/zenko-base-2.0.0-alpha.3/operator.      yaml | kubectl apply -f -
+
+#. Install XDM-Operator with:
```
Assuming "XDM Operator" is a thing... no hyphen, leave the O capitalized.

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+      /srv/scality/metalk8s-2.6.0/solutions.sh add-solution --name zenko       --solution zenko --version 2.0.0-alpha.3
+      kubectl -n zenko rollout status --timeout 10m deploy zenko-operator
+
+Deploying XDM
```
Deploy XDM

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+      }
+      EOF
+
+#. Create a yaml file for the new XDM's version with:
```
the new XDM version

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+Via the User Interface
+**********************
+
+Add the following hosts to the browser user machine ``/etc/hosts`` to resolve to 
```
to your ``etc/hosts`` file, resolving them to the MetalK8s node's IP address:

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+
+.. note::
+
+   If the machine has port 80 open, you can access the User Interface through http://ui.zenko.local.
```
```suggestion
   If port 80 is open, you can access the user interface through http://ui.zenko.local.
```

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+
+.. important::
+
+   The User Interface is limited to creating accounts and locations.
```
```suggestion
   The user interface is limited to creating accounts and locations.
```

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+Testing
+-------
+
+Via the User Interface
```
Using the GUI

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+
+   The User Interface is limited to creating accounts and locations.
+
+Via Command Line
```
Using the Command Line

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+~~~~~~~~~~~~~~
+
+- jq and curl must be installed
+- Access from the browser user machine
```
You must have browser access to the host

but... if you're doing this from the command line, why do you need browser access?

thomasj-tech replied:
I'm not sure of what that means, I asked @terrydervaux

wabernat commented on docs/docsource/installation/install/installing_xdm.rst:
```diff
@@ -0,0 +1,517 @@
+          "http://management.zenko.local/api/v1/config/${INSTANCE_ID}/endpoint" | \
+          jq '.'
+
+#. Generate the account's key with:
```
"account key" is fine.

wabernat commented on docs/docsource/operation/Upgrading_XDM.rst:
```diff
@@ -0,0 +1,28 @@
+   
+   .. code::
+      
+      curl -u '$CREDS' https://eve.devsca.com/github/scality/zenko/artifacts/builds/github:scality:zenko:staging-2.0.0.r210210230152.7c4bbdc.pre-merge.00015938/zenko-2.0.0-alpha.3.iso -o zenko-2.0.0-alpha.3.3.iso
```
This looks like a bunch of alpha and pre-release referents. When this is 2.0.0 GA, what will the targets be named?

thomasj-tech replied:
`2.0.0-beta.x` so I just changed `alpha` into `beta` where it seemed right

wabernat replied:
Use the |version| variable to replace the entire version number. Here, this should be: 
`zenko-|version|.iso `. Search the whole document and wherever you see 2.0.0-alpha.x or beta.y, replace it with |version|.

wabernat commented on docs/docsource/operation/Upgrading_XDM.rst:
```diff
@@ -0,0 +1,28 @@
+
+   .. code::
+      
+      /srv/scality/metalk8s-2.6.0/solutions.sh import --archive /home/centos/zenko-2.0.0-alpha.3.3.iso
```
The current MK8s release is 2.7.1. Ask the SME if we want to keep documenting with 2.6.0. Same with two instances below (lines 22 and 28)

thomasj-tech replied:
I asked @thomasdanan about that

wabernat replied:
You might want to ask someone on Zenko this question. It's possible that Zenko is tested on 2.6.0 and not on 2.7.1. It is also possible that it can update with no problem every time we update Zenko.

wabernat commented on docs/docsource/operation/index.rst:
```diff
@@ -15,6 +15,7 @@ such as Google Cloud Platform (GCP) or Microsoft Azure.
    :maxdepth: 2
 
    Introduction/index
+   Upgrading_XDM
```
Check to be sure this doesn't fit in with Installation. There is a Zenko Upgrade chapter over there. I'm open to moving either to the other location, but we probably shouldn't have one upgrade in Operation and another in Installation.

thomasj-tech replied:
You're right, I moved the file next to Zenko Upgrade.

wabernat commented on docs/docsource/operation/index_pdf.rst:
```diff
@@ -5,6 +5,7 @@
    :maxdepth: 2
 
    Introduction/index
+   Upgrading_XDM
```
as per previous

lucieleonard commented:
/wait

thomasj-tech commented:
/approve

wabernat commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -150,11 +150,11 @@ Deploying XDM
    .. code::
 
-      kubectl apply --namespace zenko -f /srv/scality/zenko-2.0.0-alpha.3/zenkoversion.yaml
+      kubectl apply --namespace zenko -f /srv/scality/zenko-2.0.0-beta.3/zenkoversion.yaml
```
I recommend inserting a |version| variable and setting this code sample (and subsequent ones containing a Zenko version number) as parsed literals.

thomasj-tech replied:
I've added a variable for Zenko, but the problem remains for MetalK8s. 

We are recommended to avoid writing the version number in the doc, so should I write "{{version}}" instead of MetalK8s' "2.6.0" version number?

wabernat replied:
```
.. parsed-literal::

   kubectl apply --namespace zenko -f /srv/scality/zenko-|version|/zenkoversion.yaml
```

wabernat commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -59,20 +59,20 @@ Deploy XDM Operator
+      sed "s/SOLUTION_ENV/zenko/g" /srv/scality/zenko-base-|version|-beta.3/operator.yaml | kubectl apply -f -
       kubectl -n zenko rollout status --timeout 10m deploy kubedb-operator
-      sed "s/SOLUTION_ENV/zenko/g" /srv/scality/zenko-base-2.0.0-beta.3/operator.      yaml | kubectl apply -f -
+      sed "s/SOLUTION_ENV/zenko/g" /srv/scality/zenko-base-|version|-beta.3/operator.yaml | kubectl apply -f -
```
```suggestion
      sed "s/SOLUTION_ENV/zenko/g" /srv/scality/zenko-base-|version|/operator.yaml | kubectl apply -f -
```

wabernat replied:
"-beta.3" is part of the version number. Replace these globally (carefully!)

wabernat replied:
This appears to be present in all subsequent edits. 

In Zenko development, they use a mess of alpha- and beta- tags. These never see the light of day as far as customers are concerned. If you need testability, set the version variable to "2.0.0-beta.3" (or whatever) and this will update wherever |version| is in the docs.

thomasj-tech replied:
According to the projects_common_vars.py file, "The full version, including alpha/beta/rc tags" is "release = '2.0.0'" so I removed the version suffix and let |version| where it was necessary.

thomasj-tech commented:
According to the projects_common_vars.py file, "The full version, including alpha/beta/rc tags" is "release = '2.0.0'" so I removed the version suffix and let |version| where it was necessary.

wabernat commented on docs/docsource/installation/upgrade/upgrade_xdm.rst:
```diff
@@ -7,22 +7,22 @@ Upgrade XDM
    .. parsed-literal::
       
-      curl -u '$CREDS' https://eve.devsca.com/github/scality/zenko/artifacts/builds/github:scality:zenko:staging-|version|.r210210230152.7c4bbdc.pre-merge.00015938/zenko-|version|-beta.3.iso -o zenko-|version|-beta.3.3.iso
+      curl -u '$CREDS' https://eve.devsca.com/github/scality/zenko/artifacts/builds/github:scality:zenko:staging-|version|.r210210230152.7c4bbdc.pre-merge.00015938/zenko-|version|.iso -o zenko-|version|.iso
```
You're going to have to get an engineer to spell out what "r210210230152.7c4bbdc.pre-merge.00015938" is. This may be tied to the version (like a checksum or a build number, for example) and may need to be updated in our documentation with each version (in which case, a substitution variable may be in order). On the other hand, it may be tied to the individual installation, in which case, we need to describe what this number/hash is, and where the reader can find and copy it.

thomasj-tech replied:
Yes indeed, do you know the meaning of this number @xaf-scality ? Here, |version| = 2.0.0.

miniscruff replied:
r210210230152.7c4bbdc.pre-merge.00015938 is a specific build id, my approximate guess is:
r: revision?
210210230152: timestamp probably UTC: 21-02-10 23:01:52
7c4bbdc: Most likely the git hash of the commit
pre-merge: eve build step name, we use pre-merge a lot
00015938: I believe this is the build number for eve

This changes with every eve build and tbh I wouldn't include anything that is not promoted in any of our docs as they expire after two weeks. The promoted ones have a release version followed by `-promoted`.

wabernat replied:
We'll need to spell out how the reader will access these hashes.

miniscruff replied:
Maybe @tcarmet  knows more but I wouldn't expect eve to be our preferred source of artifacts.

thomasj-tech replied:
Yes, it seems it is an internal link.

wabernat replied:
This is a packaging failure. There is no .iso in the public GitHub repository, and Eve is not public. Where are users going to find the .iso file? In earlier versions, the only engineer who had an opinion insisted that the customers build their own .iso file, and that if they didn't have that skill, they weren't clever enough to use our product. I did not find this answer satisfactory, but I do not recall how we resolved the issue. 

Please keep pushing on both development and release engineering until we get a meaningful solution.

tcarmet replied:
> Maybe @tcarmet knows more but I wouldn't expect eve to be our preferred source of artifacts.

I confirm what @miniscruff is mentioning, that link is internal, for devs only, and not meant  for the outside world.

wabernat replied:
How will the end user acquire the .iso file?

tcarmet replied:
For open source user I believe the argument is that they should build it themself, and for customers it's going to be hosted on packages.scality.com

thomasj-tech commented:
The new upgrade procedure should replace the content of upgrade-zenko.rst which is wrong for the next gen product.

wabernat commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -279,7 +279,7 @@ Deploy |product|
       kubectl wait --for condition=Ready=True --timeout 5s volume zenko-mongodb       zenko-kafka zenko-s3data zenko-redis zenko-zookeeper
 
-#. Create an |product| resource:
+#. Create a |product| resource:
```
We've reached a limit of substitution here. Either "a" or "an" is going to sound bad for one possible outcome: "an Zenko resource" or "a XDM resource".

This is fine as it is. We'll have to figure out what to do about these.

miniscruff replied:
We could substitute the a too :laughing:

miniscruff commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -0,0 +1,523 @@
+Add the following hosts to your ``etc/hosts`` file, resolving them to the 
+MetalK8s node's IP address:
+
+- management.zenko.local
+
+- keycloak.zenko.local
+
+- ui.zenko.local
+
+- s3.zenko.local
```
Shouldn't have new lines here
```suggestion
- management.zenko.local
- keycloak.zenko.local
- ui.zenko.local
- s3.zenko.local
```

miniscruff reviewed (changes requested):
Can you squash the commits to just one, no need for 12

terrydervaux commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -0,0 +1,523 @@
+      ---
+      EOF
+
+#. Create |product| volumes:
```
point to kubernetes volume creation processus, mention here that |product| will claim the following volumes with that specific storage class

lucieleonard replied:
Remove this step and replace with a link to MetalK8s Volumes creation topic

lucieleonard commented:
/wait

miniscruff commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -0,0 +1,460 @@
+Deploy |product| Operator
+-------------------------
+
+#. From your terminal, create a namespace for |product|:
```
Might be worth mentioning metal calls this an environment and not just a namespace

lucieleonard replied:
So: "Create an XDM namespace (also called environment in the Metalk8s context)"?

miniscruff replied:
No, MetalK8s environments are k8s namespaces with a couple more things related to solutions. See metal docs https://metal-k8s.readthedocs.io/en/latest/developer/architecture/solutions.html#solution-environment

thomasj-tech replied:
I simply put it like this: "From your terminal, create an environment for |product|". 
Is it too vague?

miniscruff replied:
That works for me

thomasj-tech committed:
ZENKO-3254: Fixed index toctree and some sentences

wabernat reviewed (commented):
On line 220 of install_xdm.rst, change the title to *MetalK8s Operation*

thomasj-tech committed:
ZENKO-3254: Changed doc reference name

miniscruff reviewed (approved):
It looks right to me for now, but it may change later on.

thomasj-tech committed:
ZENKO-3254: Fixed reference style

terrydervaux commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -0,0 +1,460 @@
+          -H "Content-Type: application/json" \
+          -d "${USER_PARAMS}" \
+          "http://management.zenko.local/api/v1/config/${INSTANCE_ID}/user" | \
+          jq '.'
```
Should be moved to a specific cookbook named "Create an account"

terrydervaux commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -0,0 +1,460 @@
+          -H "Content-Type: application/json" \
+          -d "${ENDPOINT_PARAMS}" \
+          "http://management.zenko.local/api/v1/config/${INSTANCE_ID}/endpoint" | \
+          jq '.'
```
Should be moved to a specific cookbook named "Create an S3 endpoint"

terrydervaux commented on docs/docsource/installation/install/install_xdm.rst:
```diff
@@ -0,0 +1,460 @@
+          -H "X-Authentication-Token: ${TOKEN}" \
+          -H "Content-Type: application/json" \
+          "http://management.zenko.local/api/v1/config/${INSTANCE_ID}/user/$      {ZENKO_ACCOUNT}/key" | \
+          jq '.'
```
Should be moved to a specific cookbook named "Generate Account Root Access Key"

terrydervaux replied:
should be updated to 

`# kubectl --kubeconfig=/etc/kubernetes/admin.conf -n zenko exec $VAULT_CONTAINER -- bash -c   "ADMIN_ACCESS_KEY_ID=${ADMIN_ACCESS_KEY_ID} ADMIN_SECRET_ACCESS_KEY=${ADMIN_SECRET_ACCESS_KEY} /vault/node_modules/vaultclient/bin/vaultclient generate-account-access-key --name=account --name account1"`

wabernat committed:
documentation/ZENKO-3254_add_installation_and_upgrade_procedures

Made changes as directed in PM comments: moved CLI commands to individual "cookbook" steps.

Consider moving to these to includes in a subsequent revision.

terrydervaux commented on docs/docsource/operation/CLI/generate_an_account_root_access_key.rst:
```diff
@@ -0,0 +1,15 @@
+.. _Generate an Account Root Access Key:
```
```suggestion
.. _Generate Account Root User Access Keys:
```

wabernat replied:
Is the point of this that when one generates an access key, one generates an access key and a secret key? If so, I'd argue for "Generate an Account Root Access Key Pair."

wabernat replied:
changed to: Generate Root User Account Access Keys

terrydervaux commented on docs/docsource/operation/CLI/generate_an_account_root_access_key.rst:
```diff
@@ -0,0 +1,15 @@
+.. _Generate an Account Root Access Key:
+
+Generate an Account Root Access Key
```
```suggestion
Generate Account Root User Access Keys
```

terrydervaux commented on docs/docsource/operation/CLI/generate_an_account_root_access_key.rst:
```diff
@@ -0,0 +1,15 @@
+Generate an Account Root Access Key
+===================================
+
+To generate an account key from the command line:
```
```suggestion
To generate account root user access keys from the command line:
```

wabernat replied:
changed to "root user account access keys"

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-3254 contains:


* `2.x`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.0.0`


Please check the `Fix Version/s` of ZENKO-3254, or the target
branch of this pull request.




*The following options are set:* **approve**

lucieleonard commented:
fix version done hellooooo

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`



The following branches will **NOT** be impacted:

* `development/0.9`
* `development/1.0`
* `development/1.1`
* `development/1.2`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`



The following branches have **NOT** changed:

* `development/0.9`
* `development/1.0`
* `development/1.1`
* `development/1.2`



Please check the status of the associated issue ZENKO-3254.

Goodbye thomasj-tech.
